### ISSUE TITLE
Seeing False Positive/Negative results from flake8 linter integration

### ISSUE THREAD

**Issue by reactive-firewall** [houndci/hound#1529]
Hi @houndci

I use and love the hound product you've built. One One pain point I've been running into is with the Flake8 integration.

FAQ:

Q: What happened exactly:
A: Hound reported a regression (see reactive-firewall/PiAP-python-tools#62 for details) on of an unchanged flake8.ini file when merging (reactive-firewall/PiAP-python-tools#61)

Q: why not fix the flake8.ini
A: because the error is only reported by Hound and not by whole CI-suite running flake8 with same config (TL;DR; see style tests from [Circle-CI](https://circleci.com/gh/reactive-firewall/PiAP-python-tools/131) and [Travis-CI](https://travis-ci.org/reactive-firewall/PiAP-python-tools/builds/371874143?utm_source=github_status&utm_medium=notification) Note: both are configured to run flake8 linter with the same config file as passed to hound)

Q: why is this considered a regression:
The previous PR passed hound and without change to hound or flake8 config files hound reports an invalid config file for flake8  (see reactive-firewall/PiAP-python-tools#62 for details).

Q: Could hound have changed, upgraded or otherwise fixed a previous false negative on the previous PR?
A: It is possible, however unlikely; TL;DR; my 20 minutes of searching through the recent PRs and closed issues yielded nothing to indicate this is the case. Further the config file is cross-checked by running the linter in multiple environments.

Q: what would you consider a fix?
A: A fix would be for the hound linter to agree with the command line linter results. 

Q: How can this be tested?
A: I'm lazy and would like someone else to suggest a non-manual way to do this. If I wasn't I could figure out how to write a wrapper for flake8 that breaks the flake8.ini config and would thus not use hound. _Instead_ I'll trust my Hound-Overlords to continue to build such an awesome product.

**Comment by gylaz** [comment]
Hi @reactive-firewall,

Thanks for the detailed write-up! This is indeed strange since nothing changed between the PRs in regards to Flake8 configuration. What version of Flake8 are you running locally and on CI? Hound is currently on `3.5.0` and I just want to make sure version mismatch is not a culprit here.

**Comment by gylaz** [comment]
It seems the culprit are the last two lines in the `exclude` section.
Should there be a comma between `build` and `os.abort()` lines? What is the purpose of `os.abort()` as the last line in the `exclude` section? It does look like that line was modified fairly recently.

**Comment by reactive-firewall** [comment]
OPPS ... `os.abort()` is meant to be in the coverage  config file (I hope that was not being run by any python like `os.path.abs(os.abort())`)
@gylaz Thanks for catching that! (missing punctuation strikes again)


While that pesky comma fixes my immediate issue (reactive-firewall/PiAP-python-tools#62) it does not explain why houndci suddenly gained the super power to lint the flake8.ini

Status FAQ:

Q: Is this fully resolved?
A: Not fully.

Q: are you unblocked?
A: yes.

Q: How satisfied with Houndci are you with the response on this issue?
A: 👍 insightful

Q: is the root cause clear?
A: no, but close enough.

Q: is the direct/'high level' cause clear?
A: yes. The lack of a comma combined with a newline caused hound to complain. TL;DR: I tested leaving in the `os.abort()` btw so no issue with code)

Q: could this be a version difference as mentioned?
A: Unsure. TL;DR; Version being the cause would be interesting ... although seems unlikely as I'm currently also using version 3.5.0 but my requirement has only been flake8>=2.5.4 for a while (indicating a false success previously). But this would need to be tested more and doesn't explain why the CI testing with the current version of Flake8 3.5.0 does not catch the config issue like hound.

Q: is this still a bug or more a lack of documentation or gap in the spec?
A:  Unsure. This is not likely a bug. either a lack of spec (i.e. what to do with multi-line exceptions) or documentation.

Q: can this issue be closed as won't fix?
A: if so desired, but not recommended due to lack of clear root cause and resolution. TL;DR; There is nothing in the Flake8 documentation to exclude the idea that this was a semantic error rather than a syntax error (as in my specific case) At a minimum this should be escalated to the Flake8 team.
Here is my humble Rational:
The result of hound to flag a missing comma leads to a cleaner more consistent config file. It is easier to say arbitrarily that multi-line exceptions are very 'un-pythonic' and thus 'style violations' than to update the code to allow them and miss this kind of corner case in the future.
Caveat: This is likely beyond the scope of the linter and not covered by [PEP8](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds)
This is still a :+1: point for hound, comma is better than mixed multiline and single line

Q: if this is selected for a fix what are you looking for hound to do:
either 
- (i) subdue the flake8 team and and get the fix up stream (unlikely)
- or (ii) just add a quick blurb in the documentation about flake via hound differs on this corner case
- or (iii) bend-over backwards to match the linter Flake8 and all of it's quarks to attain 'a foolish consistency'

**Comment by gylaz** [comment]
>  just add a quick blurb in the documentation about flake via hound differs on this corner case

This is most likely what we're going to do. Since you seem to have the most experience with this issue, any suggestions on how exactly to word this documentation blurb?

**Comment by reactive-firewall** [comment]
Q: what to say to make the devs happy?
A: something like: 
>## Note:
> At the time of writing this (May 2018) [Flake8](http://flake8.readthedocs.org/en/latest/) does not yet document the exact behavior of multi-line statements in `exclude:` stanza in the .ini file. In the spirit of consistency, HoundCI will flag uncommented lines as config errors when not strictly comma delimited.
>
> #### POOR STYLE
> insert tested failure case example of flake8.ini working config here
> #### GOOD STYLE
> insert tested example of working config here
>
and perhaps even the subtext :wink: :
> (Thanks to the [community](https://github.com/houndci/hound/issues/1529) for finding this)

Major Caveat: Hrrm, this is why I deal with computer languages rather than natural languages for a living... Not really sure on this sort of thing (all you technical writers have serious respect from me)

Q: is this copy-paste ready?
A: NOT at all! Please have a human proof-read.

Q: Where?
A: around the python docs: https://houndci.com/configuration#python

Q: Anything else?
A: yeah, the default config on that page is a broken link to http://flake8.readthedocs.io/en/latest/config.html#default

...And in case you are wondering, yes I'm currently available for hire.

**Comment by gylaz** [comment]
Thanks for the write up. We'll get this into our docs.

**Comment by gylaz** [comment]
Looks like this has been addressed and documented in http://help.houndci.com/configuration/flake8

Closing.
